packages/react-dom/src/client/ReactDOMFiberComponent.js

```diff
@@ -80,6 +93,235 @@ let normalizeHTML;
+    }
+    return '<' + normalizeNodeName(node.nodeName) + '>';
+  };
+
+  getNodeSignatureForDiff = function(node: Node, openTagOnly = false) {
+    if (node.nodeType !== ELEMENT_NODE) {
+      return getSpecialNodeSignature(node);
+    }
+    const tagName = normalizeNodeName(node.nodeName);
+    let ret = '<' + tagName;
+    if (node instanceof Element) {
+      const attrs = node.attributes;
+      const ic = attrs.length;
+      for (let i = 0; i < ic; ++i) {
+        // TODO: Should we process the value with clipStringWithEllipsis?
+        // TODO: How do we want to print quotes and other HTML special characters inside attribute quotes?
+        ret += ' ' + attrs[i].name + '="' + attrs[i].value + '"';
+      }
+    }
+    if (openTagOnly) {
+      ret += '>';
+    } else {
+      const childrenContent =
+        node instanceof Element ? node.innerHTML : node.textContent;
+      if (omittedCloseTags.hasOwnProperty(tagName)) {
+        ret += ' />';
+      } else {
+        ret +=
+          '>' +
+          clipStringWithEllipsis(childrenContent, NODE_CONTENT_CLIP_LENGTH) +
+          '</' +
+          tagName +
+          '>';
+      }
+    }
+    return ret;
+  };
+
+  const objectPrototypeToString = Object.prototype.toString;
+
+  // TODO: Reuse RESERVED_PROPS from ReactPartialRenderer.
+  const RESERVED_PROPS = {
+    children: null,
+    dangerouslySetInnerHTML: null,
+    suppressContentEditableWarning: null,
+    suppressHydrationWarning: null,
+  };
+
+  getTagWithPropsSignature = function(tag: string, props: Object) {
+    let ret = '<' + tag;
+
+    for (const propKey in props) {
+      if (!props.hasOwnProperty(propKey)) {
+        continue;
+      }
+      let propValue = props[propKey];
+      if (propValue == null) {
+        continue;
+      }
+      if (RESERVED_PROPS.hasOwnProperty(propKey)) {
+        continue;
+      }
+      let markup = null;
+      if (propKey === STYLE) {
+        markup = propKey + '={…}';
+      } else if (typeof propValue === 'function') {
+        markup = propKey + '={' + propValue.name || propValue.toString() + '}';
+      } else {
+        try {
+          markup = propKey + '={' + JSON.stringify(propValue) + '}';
+        } catch (ex) {
+          markup =
+            propKey + '={' + objectPrototypeToString.call(propValue) + '}';
+        }
+      }
+      if (markup) {
+        ret += ' ' + markup;
+      }
+    }
+
+    if (omittedCloseTags.hasOwnProperty(tag)) {
+      ret += ' />';
+    } else {
+      if (typeof props.children === 'string') {
+        ret += '>' + props.children + '</';
+      } else if (Array.isArray(props.children)) {
+        ret += '>{[';
+        const ic = props.children.length;
+        for (let i = 0; i < ic; ++i) {
+          ret += i > 0 ? ', ' : '';
+          const child = props.children[i];
+          if (typeof child === 'string') {
+            ret += "'" + child + "'";
+          } else {
+            ret += '…';
```

Comment:
Nit: can we avoid Unicode in source and use an escape?

Also are we confident ellipsis is printed properly in, say, IE11? Or when the script is served with incorrect char set? Maybe it would be better to stick to ASCII.

Reply:
Okay, yes, to simplify I'll replace with three ASCII dots.

Reply:
Done during rebase to latest `fasebook/master`.